crates/revm/src/evm_impl.rs

```diff
@@ -525,18 +570,33 @@ impl<'a, GSPEC: Spec, DB: Database, const INSPECT: bool> EVMImpl<'a, GSPEC, DB,
             self.data.db,
         ) {
             self.data.journaled_state.checkpoint_revert(checkpoint);
-            return (e, gas, Bytes::new());
+            return Err((e, gas, Bytes::new()));
         }
 
+        let contract = Box::new(Contract::new_with_context(
+            inputs.input.clone(),
+            &bytecode,
+            &inputs.context,
+        ));
+
+        Ok((gas, Box::new(bytecode), checkpoint, contract))
+    }
+
+    /// Main contract call of the EVM.
+    fn call_inner(&mut self, inputs: &mut CallInputs) -> (InstructionResult, Gas, Bytes) {
+        let res = self.prepare_call(inputs);
+        if let Err(ret) = res {
+            return ret;
+        }
+
+        let (gas, bytecode, checkpoint, contract) = res.unwrap();
+
         let ret = if is_precompile(inputs.contract, self.precompiles.len()) {
-            self.call_precompile(gas, inputs.contract, inputs.input.clone())
+            self.call_precompile(inputs, gas)
         } else if !bytecode.is_empty() {
```

Comment:
This line can be replaced with 
```suggestion
        } else if !contract.is_empty() {
```

if fn `is_empty` is not present you can make it and inside of it call `bytecode.is_empty()`